From 6639c0ddd553579b2b44dc408ac2f1809a587dca Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:00:06 -0400
Subject: [fs] btrfs: fix transaction abortion when remounting btrfs from RW to RO

Message-id: <1407179109-13714-120-git-send-email-sandeen@redhat.com>
Patchwork-id: 87078
O-Subject: [RHEL7.1 PATCH 119/422] Btrfs: fix transaction abortion when remounting btrfs from RW to RO
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit 2c21b4d733d6e50514e30ffd87110364ddda695b
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Tue Jan 14 19:42:20 2014 +0800

    Btrfs: fix transaction abortion when remounting btrfs from RW to RO

    Steps to reproduce:
     # mkfs.btrfs -f /dev/sda8
     # mount /dev/sda8 /mnt -o flushoncommit
     # dd if=/dev/zero of=/mnt/data bs=4k count=102400 &
     # mount /dev/sda8 /mnt -o remount, ro

    When remounting RW to RO, the logic is to firstly set flag
    to RO and then commit transaction, however with option
    flushoncommit enabled,we will do RO check within committing
    transaction, so we get a transaction abortion here.

    Actually,here check is wrong, we should check if FS_STATE_ERROR
    is set, fix it.

    Reported-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Suggested-by: Miao Xie <miaox@cn.fujitsu.com>
    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index b1e99ad..44bfc1e 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -8492,7 +8492,7 @@ int btrfs_start_delalloc_inodes(struct btrfs_root *root, int delay_iput)
 {
  int ret;
 
- if (root->fs_info->sb->s_flags & MS_RDONLY)
+ if (test_bit(BTRFS_FS_STATE_ERROR, &root->fs_info->fs_state))
   return -EROFS;
 
  ret = __start_delalloc_inodes(root, delay_iput);
@@ -8518,7 +8518,7 @@ int btrfs_start_delalloc_roots(struct btrfs_fs_info *fs_info, int delay_iput)
  struct list_head splice;
  int ret;
 
- if (fs_info->sb->s_flags & MS_RDONLY)
+ if (test_bit(BTRFS_FS_STATE_ERROR, &fs_info->fs_state))
   return -EROFS;
 
  INIT_LIST_HEAD(&splice);
-- 
1.7.1